User service 사용자 인증 및 조회 기능 추가 및 관련 공통 모듈 오류 수정

kym8821 opened the pull request:
## 주요 변경사항

- **사용자 인증 및 인가 기능 추가**
  - JWT 기반 로그인 및 토큰 검증 로직 구현
  - AuthTokenService와 KakaoAuthHttpClient 연동
  - 이메일 기반 사용자 조회 기능 추가
    - UserRepository에 이메일로 사용자 조회하는 메서드 추가

- **Redis를 통한 인증 정보 관리**
  - 액세스/리프레시 토큰 저장 및 검증 로직 구현

- **인증 관련 공통 모듈 추가(common) 및 공통 모듈 빌드 오류 수정**
  - jwt 토큰 관리 기능과 상수값들을 공통 모듈로 분리
  - 루트 모듈 빌드 / 테스트 코드 실행 시 공통 모듈로 인해 발생하는 에러 수정 (main 클래스 관련 에러)

- **테스트 코드 작성 및 설정 개선**
  - 인증 관련 단위/통합 테스트 추가
  - 테스트용 설정 파일을 Git submodule로 분리 관리

## 기타
- 테스트 환경에서는 `application-test.yml`을 별도로 사용
- KakaoAuthHttpClient는 테스트 시 Mock 처리
- 추후 redis 관련 설정을 공통 모듈로 분리하는 것도 고려할 필요 있음

kym8821 committed:
[feat] add common module
- jwt 토큰 인증에 필요한 기능을 담는 클래스 추가 (JwtUtils)
- jwt 인증 관련 상수값을 담는 클래스 추가 (AuthProperty)

kym8821 committed:
[feat] user service 생성
- settings.gradle에 common, user-service 추가
- build.gradle에 의존성 추가
- user service의 main 클래스 추가

kym8821 committed:
[feat] entity와 repository 생성
- 사용자 entity (User) 추가
- 구독 관계를 나타내기 위한 엔티티 추가 (Subscription)
- 사용자 성별을 나타내기 위한 열거형 객체 추가 (Gender)
- User의 JPA Repository 레이어 추가 (UserRepository)

kym8821 committed:
[feat] 카카오 로그인 기능 초안 생성
- KakaoAuthHttpClient : 카카오 로그인 기능 초안 (기능 추가 필요)
- KakaoUserInfo : 카카오 사용자 정보 dto (추후 수정 필요)

kym8821 committed:
[feat] 사용자 인증 기능 서비스 레이어 초안 생성
- UserAuthResponse : 사용자 인증 정보 저장 (redis 연동 및 카카오 로그인 구현 후 수정 필요)
- UserAuthService : 사용자 인증 서비스 레이어  (카카오 로그인 구현 후 수정 필요)

kym8821 committed:
[fix] common 모듈 빌드 문제 해결
- jar.enabled=true 설정
- main 클래스 추가를 통해 main 클래스를 못찾는 문제 해결

kym8821 committed:
[chore] update submodule
- user-service 설정 정보 변경

kym8821 committed:
[feat] 카카오 로그인 기능 추가
- KakaoAuthHttpClient : 인가 코드로 사용자 인증 및 정보 조회하는 기능 추가
- KakaoUserInfo : Entity로 변환 기능, 카카오가 제공한 사용자 정보 변환 기능 추가
- KakaoRequestDto : access token 요청의 요청 본문 클래스
- KakaoResponseDto : access token 응답, 사용자 정보 응답의 응답 본문 클래스

kym8821 committed:
[fix] 사용자 로그인 기능을 webclient 방식으로 변경
- KakaoUserInfo 객체가 아닌 Mono<KakaoUserInfo>를 반환하도록 수정

kym8821 committed:
[chore] 설정 파일 서브모듈 update

kym8821 committed:
[fix] 공통 모듈 빌드 에러 수정
- common 모듈 main 클래스 삭제
- bootjar enabled = false, jar enabled = true 설정 추가

kym8821 committed:
[feat] redis와 test 코드 작성용 의존성, 작업 추가
- redis 의존성 추가
- 테스트용 의존성 추가 (starter-test, h2, java-faker)
- 테스트용 설정 파일 서브모듈로 분리

kym8821 committed:
[feat] redis 설정 추가 및 저장소 접근 기능 추가
- RedisTemplate, RedisConnectionFactory 설정용 빈 추가 (RedisConfig)
- Redis 데이터 접근용 빈 추가 (RedisSingleDataService)
- 인증 및 인가 토큰 접근용 빈 추가 (TokenService)

kym8821 committed:
[feat] 인증 및 인가 비즈니스 로직 추가
- 트랜잭션이 적용된 카카오 로그인 기능 추가
- 로그아웃 기능 추가
- 토큰 재발급 기능 추가
- 카카오 사용자 정보에 birth 추가 (KakaoUserInfo)
- 성별 열거형의 명칭 변경 (MALE, FEMAIL -> male, female)

kym8821 committed:
[feat] 사용자 인증 및 인가 컨트롤러 및 DTO 생성
- 사용자 인증 및 인가 컨트롤러 추가 ( UserAuthController )
- 사용자 로그인 DTO 추가 (UserAuthRequest )

kym8821 committed:
[feat] 사용자 조회 기능 추가
- 사용자 조회 컨트롤러 추가 (UserRetrieveController)
- 사용자 조회 비즈니스 로직 추가 (UserRetrieveService)
- 사용자 조회 응답 DTO 추가 (UserResponseDto)

kym8821 committed:
[test] 사용자 테스트 데이터 추가 및 비즈니스 로직 테스트 코드 작성
- 사용자 테스트 데이터 추가 기능 (UserTestDataGenerator)
- 사용자 인증 비즈니스 로직 성공 테스트 (UserAuthServiceTest)
- 사용자 조회 비즈니스 로직 성공 테스트 (UserRetrieveServiceTest)

kym8821 committed:
[test] 토큰 관리 객체의 테스트 코드 작성 (JwtUtilsTest)

klaqwe1 commented on common/src/main/java/com/theme/auth/JwtUtils.java:
```diff
@@ -0,0 +1,93 @@
+            }
+            return authorization.substring(AuthProperty.ACCESS_TOKEN_PREFIX.length());
+        }catch (Exception e) {
+            return null;
```
예외 처리 시 토큰확인 실패 log 를 남기면 어떨 까 싶습니다

kym8821 replied:
이 부분 전적으로 동의합니다. 실패 로그 남기는 부분 추가하겠습니다.

klaqwe1 commented on user-service/src/main/java/com/theme/redis/RedisConfig.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Bean
+    public RedisConnectionFactory redisConnectionFactory() {
+        return new LettuceConnectionFactory(host, Integer.parseInt(port));
```
port를 string에서 int로 변환하기보다 직접 int로 받는 것은 어떤가요?

kym8821 replied:
그러네요. 바로 설정 값을 int로 받앙올 수 있는 방법이 있네요. 현재 발생한 충돌 해결하면서 현재 말씀해주신 부분 수정 후 merge 하겠습니다.

klaqwe1 reviewed (commented):
### **_고생 많으셨습니다!!_**

메소드 이름이 명확하고, 기능별로 세분화 되어있어 코드 리뷰하기 좋았습니다.
다만, **주석**으로 기능을 명시한다면 팀원들이 더욱 쉽게 코드를 리뷰할 수 있을 것 같습니다!

그리고 JWT 토큰 처리, 인증 코드가 엄청났습니다... JWT의 장인이 되신 것 같아 감탄했습니다.

용민님과 같은 개발자와 함께 작업할 수 있어서 정말 **행운**이라고 생각합니다! 고생 많으셨습니다!

kym8821 committed:
[merge] merge user-service with dev

kym8821 committed:
[fix] 에러 로깅 기능 추가 및 설정 파일 데이터 불러오는 방식 변경
- JwtUtils : 에러 catch 시 에러를 로깅하도록 수정
- RedisConfig : 설정 파일의 port를 int값으로 불러오도록 수정
